### PR TITLE
rename subfolder mbed-util to core-util

### PR DESCRIPTION
**Pull request by rgrover**
@0xc0170 @bogdanm @bremoran

### CODE REVIEW

File: core-util/atomic_ops.h [hunk 1]
```diff
@@ -99,6 +99,17 @@ bool atomic_cas(T *ptr, T *expectedCurrentValue, T desiredValue)
     return rc;
 }
 
+/* For ARMv7-M and above, we use the load/store-exclusive instructions to
+ * implement atomic_cas, so we provide three specialization corresponding to the
+ * byte, half-word, and word variants of the instructions. As long as the following
+ * declarations are available, the compiler will find the proper call targets. */
+#if (__CORTEX_M >= 0x03)
```
Comment (bogdanm) [inline core-util/atomic_ops.h]: This change looks to be more than just a rename. Where is this coming from?

File: core-util/sbrk.h [hunk 2]
```diff
@@ -71,22 +75,32 @@
...
 
 #ifdef __ARMCC_VERSION
-	#define __mbed_sbrk_start (Image$$ARM_LIB_HEAP$$Base)
-	#define __mbed_krbs_start (Image$$ARM_LIB_HEAP$$ZI$$Limit)
-	#define __heap_size (Image$$ARM_LIB_HEAP$$ZI$$Length)
-	extern unsigned int __mbed_sbrk_start;
-	extern unsigned int __mbed_krbs_start;
-	extern unsigned int __heap_size;
-	#pragma import(__use_two_region_memory)
+    #define MBED_SBRK_START &(Image$$ARM_LIB_HEAP$$Base)
```
Comment (bogdanm) [inline core-util/sbrk.h]: Same question.

File: core-util/sbrk.h [hunk 3]
```diff
@@ -71,22 +75,32 @@
...
+    extern uint32_t __mbed_sbrk_start;
+    extern uint32_t __mbed_krbs_start;
+    #define MBED_SBRK_START &__mbed_sbrk_start
+    #define MBED_KRBS_START &__mbed_krbs_start
+#else
+    extern void *__mbed_sbrk_start;
+    extern void *__mbed_krbs_start;
+    #define MBED_SBRK_START __mbed_sbrk_start
+    #define MBED_KRBS_START __mbed_krbs_start
+#endif
```
Comment (bogdanm) [inline core-util/sbrk.h]: I can see these changes here, but no corresponding changes in `sbrk.cpp`, is this right?

**Comment by rgrover** [comment]
tested against mbed-drivers using frdm-k64f-gcc. All previously passing tests continue to pass.

**Comment by 0xc0170** [comment]
Is this just rename? Github shows some changes in files within the same commit :confused:
Can we split them into separate PR?

File: core-util/PoolAllocator.h [hunk 4]
```diff
@@ -45,12 +45,12 @@ class PoolAllocator {
                to satisfy the 'alignment' argument)
       * @param alignment allocation alignment in bytes (must be a power of 2, at least 4)
       */
-    PoolAllocator(void *start, size_t elements, size_t element_size, unsigned alignment = MBED_UTIL_POOL_ALLOC_DEFAULT_ALIGN);
+    PoolAllocator(void *start, size_t elements, size_t element_size, size_t alignment = MBED_UTIL_POOL_ALLOC_DEFAULT_ALIGN);
```
Comment (bogdanm) [inline core-util/PoolAllocator.h]: Not sure why you changed this type, but in any case, there are other alignments declared as `unsigned` in this module. So you should change all of them or none of them (my advice would be to leave them as `unsigned` unless there's a really good reason to change the type).

**Comment by rgrover** [comment]
sorry. it seems some changes from my work for the POSIX backend have entered into these PRs. I'll filter them out. The renames made it difficult to compare.

**Comment by 0xc0170** [comment]
what about namespace rename (mbed::util -> core::util) ?

**Comment by rgrover** [comment]
I'm happy to rename namespace mbed::util to core::util. Please object if I shouldn't.

**Comment by yogpan01** [comment]
renaming namespace will cause constant changes to modules which are using for example mbed-client-example will break , also mbed-mesh-api will break.
Please re-consider these before you go ahead and change namespacing

**Comment by bogdanm** [comment]
> what about namespace rename (mbed::util -> core::util) ?

I was in vacation for too long it seems :) When did this happen?

**Comment by rgrover** [comment]
@bogdanm  mbed::util -> core::util hasn't happened yet. will be happening shortly. (today?)

**Comment by bogdanm** [comment]
> @bogdanm mbed::util -> core::util hasn't happened yet. will happen shortly. (today?)

Sorry @rgrover, I wasn't clear. My question was "why do we want to change this?"

**Comment by rgrover** [comment]
@bogdanm it would seem a logical fallout of renaming everything as core-util. But perhaps it isn't necessary.

**Comment by 0xc0170** [comment]
@bogdanm I am asking if all this rename should also impact namespace in this module to follow, or we are happy with mbed::util for future, as below.

@yogpan01 This PR is breaking change already (we should create a label in our repositories to indicate the intention (breaking change) (not only if you used the feature "extraInclude" (mbed-util/header.h)

**Comment by bremoran** [comment]
@0xc0170 @yogpan01: This is what sem-ver is for.  This will be a minor version bump.  All new development will continue in the new minor version.  All modules depending on core-util will need to be updated to build cleanly before adopting the new version.

**Comment by rgrover** [comment]
please review
@0xc0170 @bogdanm @bremoran

**Comment by bremoran** [comment]
We should see if we can get Jenkins to cover this repo, given all the changes that are active.

**Comment by bogdanm** [comment]
+1 @bremoran

**Comment by 0xc0170** [comment]
@rgrover Can you please update also https://github.com/rgrover/core-util/blob/master/module.json ? I noticed there is still old name used.

**Comment by rgrover** [comment]
@0xc0170 please review

File: core-util/Array.h [hunk 5]
```diff
@@ -20,9 +20,9 @@
 
 #include <stddef.h>
 #include <stdint.h>
-#include "mbed-util/CriticalSectionLock.h"
-#include "mbed-util/PoolAllocator.h"
-#include "mbed-util/mbed-util.h"
+#include "core-util/CriticalSectionLock.h"
+#include "core-util/PoolAllocator.h"
+#include "core-util/mbed-util.h"
```
Comment (bogdanm) [inline core-util/Array.h]: Since we're doing this, shouldn't `mbed-util.h` be renamed to `core-util.h`?

Comment (bogdanm) [inline core-util/Array.h]: ...and all the `MBED_UTIL_` stuff inside the `mbed-util.h`. Renames are a lot trickier than they look.

**Comment by rgrover** [comment]
@bogdanm please review

File: source/mbed-util_mbed.c [hunk 6]
```diff
@@ -21,12 +21,12 @@
 #include <stdarg.h>
 #include "device.h"
 #include "mbed/mbed_interface.h"
-#include "mbed-util/mbed-util.h"
+#include "core-util/core-util.h"
```
Comment (bogdanm) [inline source/mbed-util_mbed.c]: I think the last thing left to do here is to rename `mbed-util_mbed.c` to `core-util_mbed.c` (or even `core-util_mbed_os.c`)

**Comment by rgrover** [comment]
@bogdanm

**Comment by bremoran** [comment]
The loss of Jenkins earlier today meant that we haven't gotten the build triggers for these changes.  I've manually triggered a build.

**Comment by bogdanm** [comment]
LGTM. The reported build failure should be fixed by a separate PR to MINAR, so I'm going to assume that you compiled locally and everything was fine.

**Comment by rgrover** [comment]
I've compiled locally and run tests from mbed-drivers

**Comment by bogdanm** [comment]
+1 then

**Comment by bremoran** [comment]
+1

**Comment by bogdanm** [comment]
Published new minor version `0.2.0`

**Pull request merged**